Python/specialize.c

```diff
@@ -1195,3 +1199,33 @@ _Py_Specialize_BinaryAdd(PyObject *left, PyObject *right, _Py_CODEUNIT *instr)
     assert(!PyErr_Occurred());
     return 0;
 }
+
+int
+_Py_Specialize_BinaryMultiply(PyObject *left, PyObject *right, _Py_CODEUNIT *instr)
+{
+    PyTypeObject *left_type = Py_TYPE(left);
+    if (left_type != Py_TYPE(right)) {
+        SPECIALIZATION_FAIL(BINARY_MULTIPLY, SPEC_FAIL_DIFFERENT_TYPES);
+        goto fail;
+    }
+    if (left_type == &PyLong_Type) {
```

Comment:
Ditto above about using ``Py_IS_TYPE``. Below for ``left_type == &PyFloat_Type`` too please.